tests/handling/daemons/test_daemon_errors.py

```diff
@@ -0,0 +1,181 @@
+import logging
+
+import kopf
+from kopf.storage.finalizers import FINALIZER
+
+
+async def test_daemon_stopped_on_permanent_error(
+        registry, resource, dummy, manual_time,
+        caplog, assert_logs, k8s_mocked, simulate_cycle):
+    caplog.set_level(logging.DEBUG)
+
+    @kopf.daemon(resource.group, resource.version, resource.plural, registry=registry, id='fn',
+                 backoff=0.01)
+    async def fn(**kwargs):
+        dummy.mock()
+        dummy.kwargs = kwargs
+        dummy.steps['called'].set()
+        raise kopf.PermanentError("boo!")
+
+    event_object = {'metadata': {'finalizers': [FINALIZER]}}
+    await simulate_cycle(event_object)
+
+    await dummy.steps['called'].wait()
+    await dummy.wait_for_daemon_done()
+
+    assert dummy.mock.call_count == 1
+    assert k8s_mocked.patch_obj.call_count == 0
+    assert k8s_mocked.sleep_or_wait.call_count == 1  # one for each retry
+    assert k8s_mocked.sleep_or_wait.call_args_list[0][0][0] is None
```

Comment:
This is not readable. Maybe `from unittest.mock import call` and
```
assert ...call_args_list == [call(*args, **kwargs)]
```
is better. Applies to the other stuff also. If tests breaks, then there might be obscure out of index exceptions.

Reply:
Breaking the test in that case is intended. The first index is "secured" by a preceding check for the number of calls.

But you are right, making tests more readable is a good idea — but should be done for all the tests. Now, the tests in overall are in a clumsy state, grown over the year, and they were never refactored. Some of them are duplicating each other, some aspects of the code are just missing.

I've created #338 to collect the ideas on the tests, and added this one there.